### PR TITLE
Implemented Mining to Proxy

### PR DESCRIPTION
**Pull request by Djadih**
Closes https://github.com/dominant-strategies/go-quai/issues/642

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -56,6 +56,8 @@ github.com/Azure/go-autorest/tracing v0.5.0/go.mod h1:r/s2XiOKccPW3HrqB+W0TQzfbt
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
 github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802/go.mod h1:IVnqGOEym/WlBOVXweHU+Q+/VP0lqqI8lqeDx9IjBqo=
 github.com/DATA-DOG/go-sqlmock v1.3.3/go.mod h1:f/Ixk793poVmq4qj/V1dPUg2JEAKC73Q5eFN3EC/SaM=
+github.com/INFURA/go-ethlibs v0.0.0-20230210163729-fc6ca4235802 h1:GGU/jUNUUkGbnnqi7Eoq8L9qEDG+GbW5Z/sCuYFPK9U=
```
Comment (0xalank) [inline go.sum:59]: Which variables are we using from go-ethlibs?

File: main.go [hunk 2]
```diff
@@ -118,35 +128,51 @@ func main() {
...
-func (m *Miner) client(ctx int) *ethclient.Client {return m.sliceClients[ctx]}
-
 // subscribePendingHeader subscribes to the head of the mining nodes in order to pass
 // the most up to date block to the miner within the manager.
-func (m *Miner) subscribePendingHeader() {
-	if _, err := m.client(common.ZONE_CTX).SubscribePendingHeader(context.Background(), m.updateCh); err != nil {
-		log.Fatal("Failed to subscribe to pending header events", err)
+func (m *Miner) subscribePendingHeader() error {
+	address := "0x4a06cbb083ce0ae76aa393db8acfced7552b7218"
+	password := "password"
```
Comment (0xalank) [inline main.go]: Make this an environment variable.

File: main.go [hunk 3]
```diff
@@ -250,6 +276,9 @@ func (m *Miner) resultLoop() error {
 	for {
 		select {
 		case header := <-m.resultCh:
+			// Miner only needs to be aware of the proxy (stored at zone context).
+			go m.sendMinedHeader(common.ZONE_CTX, header)
```
Comment (0xalank) [inline main.go]: If the miner only needs to be aware of the proxy do we even need slice clients?

File: main.go [hunk 4]
```diff
@@ -262,25 +291,31 @@ func (m *Miner) resultLoop() error {
...
-	if err != nil {
-		fmt.Println("error submitting block: ", err)
+	retryDelay := 1 // Start retry at 1 second
+	for {
+		header_req, err := jsonrpc.MakeRequest(0, "quai_receiveMinedHeader", quaiclient.RPCMarshalHeader(header))
+		if err != nil {
+			log.Fatalf("Could not send mined header: %v", err)
+		}
+
+		err = m.sliceClients[ctx].SendTCPRequest(*header_req)
```
Comment (0xalank) [inline main.go]: Since we're only sending blocks to the proxy and backdooring the use of `sliceClients` this PR should just remove `sliceClients` entirely in favor of a proxyConnection.

Comment (0xalank) [reply main.go]: That's of course if we do not want to have optionality to mine directly on the node or the proxy.

**Review by 0xalank — DISMISSED** [review]

File: main.go (hunk 2 above)
Comment (Djadih) [reply main.go]: Totally, this was just for development but left it in there.

File: example.yaml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+# Config for node URLs
+Location: [1,2]
```
Comment (wizeguyy) [inline example.yaml]: This file is a duplicate of `config/config.yaml.dist`. Please remove this, and add a note to the README to use `config.yaml.dist` as a distributed example config.

File: example.yaml [hunk 6]
```diff
@@ -1,9 +1,10 @@
...
+Location: []
 Auto: false
 Mine: false
 Optimize: false
 OptimizeTimer: 10
 PrimeURL: "ws://127.0.0.1:8547"
 RegionURLs: ["ws://127.0.0.1:8579", "ws://127.0.0.1:8581", "ws://127.0.0.1:8583"]
-ZoneURLs: [["ws://127.0.0.1:8611", "ws://127.0.0.1:8643", "ws://127.0.0.1:8675"], ["ws://127.0.0.1:8613", "ws://127.0.0.1:8645", "ws://127.0.0.1:8677"], ["ws://127.0.0.1:8615", "ws://127.0.0.1:8647", "ws://127.0.0.1:8679"] ]
\ No newline at end of file
+# ZoneURLs: [["http://127.0.0.1:8611", "ws://127.0.0.1:8643", "ws://127.0.0.1:8675"], ["ws://127.0.0.1:8613", "ws://127.0.0.1:8645", "ws://127.0.0.1:8677"], ["ws://127.0.0.1:8615", "ws://127.0.0.1:8647", "ws://127.0.0.1:8679"] ]
```
Comment (wizeguyy) [inline example.yaml]: Same comment as previous commit about not needing `example.yaml`

File: go.mod [hunk 7]
```diff
@@ -2,9 +2,15 @@ module github.com/dominant-strategies/quai-cpu-miner
 
 go 1.19
 
+replace github.com/dominant-strategies/go-quai => ../../djadih/go-quai
+
+replace github.com/dominant-strategies/go-quai-stratum => ../../djadih/go-quai-stratum
```
Comment (wizeguyy) [inline go.mod]: We should not commit your `replace`ments. These are just for your local development/testing

File: main.go [hunk 8]
```diff
@@ -118,35 +128,51 @@ func main() {
...
 // subscribePendingHeader subscribes to the head of the mining nodes in order to pass
 // the most up to date block to the miner within the manager.
-func (m *Miner) subscribePendingHeader() {
-	if _, err := m.client(common.ZONE_CTX).SubscribePendingHeader(context.Background(), m.updateCh); err != nil {
-		log.Fatal("Failed to subscribe to pending header events", err)
+func (m *Miner) subscribePendingHeader() error {
+	address := "0x4a06cbb083ce0ae76aa393db8acfced7552b7218"
+	password := "password"
+
+	msg, err := jsonrpc.MakeRequest(1, "quai_submitLogin", address, password)
```
Comment (wizeguyy) [inline main.go:151]: Does this mean our node API has authentication now?

File: main.go [hunk 9]
```diff
@@ -118,35 +128,51 @@ func main() {
...
 		header:         types.EmptyHeader(),
-		updateCh:      make(chan *types.Header, resultQueueSize),
+		updateCh:       make(chan *types.Header, resultQueueSize),
 		resultCh:       make(chan *types.Header, resultQueueSize),
-		previousNumber: [common.HierarchyDepth]uint64{0,0,0},
+		previousNumber: [common.HierarchyDepth]uint64{0, 0, 0},
 	}
 	log.Println("Starting Quai cpu miner in location ", config.Location)
-	m.fetchPendingHeader()
+	go m.startListener()
```
Comment (wizeguyy) [inline main.go]: Will the other go funcs fail if the listener isn't up yet?

File: main.go (hunk 4 above)
Comment (wizeguyy) [reply main.go]: I think we should still allow the miner to connect directly to the node. But, IMO these two behaviors should be distinct and clearly separated.

So `m.sliceClients` should only be setup and used if the miner is configured to connect to a slice.
Conversely, if connecting to a proxy, there should be a distinct `m.proxyClient` that the miner uses.

Thoughts about that ^?

**Review by wizeguyy — CHANGES_REQUESTED** [review]
I have several comments, but the two big requests are:
* don't commit the different YAMLs. Just provide one example YAML as `config.yaml.dist`
    * either add a note in the readme to manually copy this YAML, or add logic to the miner (or makefile) to copy it automatically if one doesn't exist
* distinctly separate slice clients from proxy clients

File: go.sum (hunk 1 above)
Comment (Djadih) [reply go.sum:59]: We're importing this library: https://pkg.go.dev/github.com/INFURA/go-ethlibs/jsonrpc#Response
It's very nice for dealing with the type of JSON RPCs that include arrays in the "Param" field by default, which is what go-quai and geth use extensively.

Incidentally, there are some improvements I'd like to see/implement and I've had an issue up in the proxy to address them. https://github.com/dominant-strategies/go-quai-stratum/issues/7

File: main.go (hunk 8 above)
Comment (gameofpointers) [inline main.go]: what is this 1 here. maybe make it a constant

File: main.go [hunk 10]
```diff
@@ -118,35 +128,51 @@ func main() {
...
 }
 
 // PendingBlocks gets the latest block when we have received a new pending header. This will get the receipts,
 // transactions, and uncles to be stored during mining.
+// As written, this will only run once at startup.
 func (m *Miner) fetchPendingHeader() {
 	retryDelay := 1 // Start retry at 1 second
 	for {
-		header, err := m.client(common.ZONE_CTX).GetPendingHeader(context.Background())
+		msg, err := jsonrpc.MakeRequest(1, "quai_getPendingHeader", nil)
```
Comment (gameofpointers) [inline main.go]: same as above

File: main.go (hunk 4 above)
Comment (gameofpointers) [reply main.go]: I agree with wizeguyy. For local testing its unnecessary to run proxy every time we want to test something. Ideally I would like a flag in config to select what I am doing.

File: example.yaml (hunk 6 above)
Comment (Djadih) [reply example.yaml]: Removed

File: go.mod (hunk 7 above)
Comment (Djadih) [reply go.mod]: Good catch, removed

File: go.sum (hunk 1 above)
Comment (Djadih) [reply go.sum:59]: MIT license confirmed. Verbally acknowledged.

File: main.go (hunk 9 above)
Comment (Djadih) [reply main.go]: Nope, confirmed by commenting out this line. The miner just won't be able to read data coming in on the TCP connection.

File: main.go (hunk 2 above)
Comment (Djadih) [reply main.go]: Done

File: config/config.yaml.dist [hunk 11]
```diff
@@ -1,3 +1,11 @@
-PrimeURL: "ws://127.0.0.1:8547"
-RegionURLs: ["ws://127.0.0.1:8579", "ws://127.0.0.1:8581", "ws://127.0.0.1:8583"]
-ZoneURLs: [["ws://127.0.0.1:8611", "ws://127.0.0.1:8643", "ws://127.0.0.1:8675"], ["ws://127.0.0.1:8613", "ws://127.0.0.1:8645", "ws://127.0.0.1:8677"], ["ws://127.0.0.1:8615", "ws://127.0.0.1:8647", "ws://127.0.0.1:8679"] ]
\ No newline at end of file
+# Config for node URLs
+Location: []
+Auto: false
```
Comment (wizeguyy) [inline config/config.yaml.dist]: Please only add viable options to the config. e.g. `Auto`, `Mine`, `Optimize`, `OptimizeTimer`, etc have no affect on CPU miner behavior. These and other such options shouldn't be included in this list

Comment (wizeguyy) [reply config/config.yaml.dist]: I think the only options used today are:
```
# Connection details to Quai nodes
PrimeURL: "ws://127.0.0.1:8547"
RegionURLs: ["ws://127.0.0.1:8579", "ws://127.0.0.1:8581", "ws://127.0.0.1:8583"]
ZoneURLs: [["ws://127.0.0.1:8611", "ws://127.0.0.1:8643", "ws://127.0.0.1:8675"], ["ws://127.0.0.1:8613", "ws://127.0.0.1:8645", "ws://127.0.0.1:8677"], ["ws://127.0.0.1:8615", "ws://127.0.0.1:8647", "ws://127.0.0.1:8679"] ]

# Proxy credentials
AccountAddress: "0x4a08cba083ce0ae76aa393db8acfced7552b7218"
Password: "password"
```
right?

File: main.go (hunk 3 above)
Comment (Djadih) [reply main.go]: Took this into account and re-added solo mining compatibility. In that case, slice clients are needed. Also, when proxy is enabled, we don't store slice clients but only the single proxyclient.

File: main.go (hunk 4 above)
Comment (Djadih) [reply main.go]: Enabled solo mining directly with the node. When proxy is enabled, no sliceclients are created and only the single proxyclient is created.

File: example.yaml (hunk 5 above)
Comment (Djadih) [reply example.yaml]: Removed both other yaml files, kept only the example.dist.

File: main.go (hunk 8 above)
Comment (Djadih) [reply main.go]: This is supposed to represent the ID for the json rpc request. Created a function to increment latestID and put that here.

File: main.go (hunk 10 above)
Comment (Djadih) [reply main.go]: Responded above

File: config/config.yaml.dist (hunk 11 above)
Comment (Djadih) [reply config/config.yaml.dist]: Correct, updated in the new version.

**Review by Djadih — COMMENTED** [review]
Updated based on comments and simplified some of the objects that were previously placed in go-quai.

File: main.go [hunk 12]
```diff
@@ -34,27 +40,53 @@ type Miner struct {
...
+	// Tracks the latest JSON RPC ID to send to the proxy or node.
+	latestId	uint64
 }
 
-// Clients for RPC connection to the Prime, region, & zone node belonging to the
+// Clients for RPC connection to the Prime, region, & zone ports belonging to the
 // slice we are actively mining
 type SliceClients [common.HierarchyDepth]*ethclient.Client
 
+// getNodeClients takes in a config and retrieves the Prime, Region, and Zone client
```
Comment (wizeguyy) [inline main.go]: (nit) stale comment

File: main.go [hunk 13]
```diff
@@ -34,27 +40,53 @@ type Miner struct {
...
+// that is used for mining in a slice.
+func connectToProxy(config util.Config) *util.MinerSession {
+	// loc := config.Location
+	zoneConnected := false
+	var client *util.MinerSession
+	var err error
+	for !zoneConnected {
```
Comment (wizeguyy) [inline main.go]: (nit) `zoneConnected` -> `proxyConnected` or just `connected`?

File: main.go [hunk 14]
```diff
@@ -93,6 +125,10 @@ func connectToSlice(config util.Config) SliceClients {
 	return clients
 }
 
+func init() {
+	log.SetFlags(log.LstdFlags | log.Lshortfile)
```
Comment (wizeguyy) [inline main.go:127]: Good change, could you put it in its own commit with an explanation?

File: main.go [hunk 15]
```diff
@@ -116,37 +152,87 @@ func main() {
...
-		previousNumber: [common.HierarchyDepth]uint64{0,0,0},
+		previousNumber: [common.HierarchyDepth]uint64{0, 0, 0},
 	}
 	log.Println("Starting Quai cpu miner in location ", config.Location)
-	m.fetchPendingHeader()
-	go m.subscribePendingHeader()
-	go m.resultLoop()
+	if config.Proxy {
+		m.proxyClient = connectToProxy(config)
+		go m.fetchPendingHeader()
```
Comment (wizeguyy) [inline main.go]: IIUC `fetchPendingHeader()` and `fetchNodeHeader()` are effectlively the same method, just with a different backend (node vs proxy). Maybe rename the first to `fetchProxyHeader()`? Or combine the two with an arg to select which backend?

Comment (wizeguyy) [reply main.go]: Same comment for `subscribeProxy()` and `subscribePendingHeader()`

File: main.go [hunk 16]
```diff
@@ -246,11 +332,17 @@ func (m *Miner) hashratePrinter() {
...
-func (m *Miner) resultLoop() error {
+func (m *Miner) resultLoop(config util.Config) {
 	for {
 		select {
 		case header := <-m.resultCh:
 			order, err := m.GetDifficultyOrder(header)
+			if !config.Proxy {
+				go m.sendMinedHeaderNodes(order, header)
+			} else {
+				// Proxy miner only needs to send to the proxy (stored at zone context).
```
Comment (wizeguyy) [inline main.go:349]: (nit) again, could use some function renames here for consistency

File: Makefile [hunk 17]
```diff
@@ -6,6 +6,15 @@ GOBIN = ./build/bin
 GO ?= latest
 GORUN = env GO111MODULE=on go run
 
+clean:
+	env GO111MODULE=on go clean -cache
+	rm -fr build/_workspace/pkg/ $(GOBIN)/*
+
+debug:
```
Comment (wizeguyy) [inline Makefile:13]: These changes are also a candidate for an independent commit

File: main.go (hunk 12 above)
Comment (Djadih) [reply main.go]: Removed

File: main.go (hunk 13 above)
Comment (Djadih) [reply main.go]: proxyConnected

File: main.go (hunk 15 above)
Comment (Djadih) [reply main.go]: Done

File: main.go (hunk 16 above)
Comment (Djadih) [reply main.go:349]: Done

File: main.go [hunk 18]
```diff
@@ -246,11 +332,17 @@ func (m *Miner) hashratePrinter() {
...
 
 // resultLoop takes in the result and passes to the proper channels for receiving.
-func (m *Miner) resultLoop() error {
+func (m *Miner) resultLoop(config util.Config) {
 	for {
 		select {
 		case header := <-m.resultCh:
 			order, err := m.GetDifficultyOrder(header)
+			if !config.Proxy {
+				go m.sendMinedHeaderNodes(order, header)
```
Comment (Djadih) [inline main.go]: Addressed a verbal comment that indicated preference for calculating and using order in the same place.
Simplified sendMinedHeaderNodes to only send to one destination, and the for loop calls that function multiple times as needed.

File: main.go [hunk 19]
```diff
@@ -246,11 +332,17 @@ func (m *Miner) hashratePrinter() {
 }
 
 // resultLoop takes in the result and passes to the proper channels for receiving.
-func (m *Miner) resultLoop() error {
+func (m *Miner) resultLoop(config util.Config) {
 	for {
 		select {
 		case header := <-m.resultCh:
 			order, err := m.GetDifficultyOrder(header)
```
Comment (Djadih) [inline main.go:333]: Improved error handling when this function returns with an error. Previously, it would have still tried to send the mined header. Now it will stop execution on this header and wait for a new one to arrive in the channel.

**Review by Djadih — COMMENTED** [review]
Several improvements, some outside of the identified comments.

File: Makefile (hunk 17 above)
Comment (Djadih) [reply Makefile:13]: Done

File: main.go (hunk 14 above)
Comment (Djadih) [reply main.go:127]: Done

**Review by Djadih — COMMENTED** [review]
Reorganized commits to better segregate features.

**Review by Djadih — COMMENTED** [review]
Combined some commits, enabled location flag in config, made lots of improvements to the readme.

**Review by wizeguyy — DISMISSED** [review]

File: README.md [hunk 20]
```diff
@@ -0,0 +1,69 @@
+# Quai CPU Miner
+Official Golang implementation of Quai's CPU Miner.
+
+# Building the source
+For prerequisites and detailed build instructions please read the Installation Instructions.
+
+Building quai-cpu-miner requires both a Go (version 1.19 or later) and a C compiler. You can install them using your favourite package manager. Once the dependencies are installed, run
+
+## Build via Makefile
+make quai-cpu-miner
```
Comment (gameofpointers) [inline README.md:11]: can we put this under ``` ```?

File: README.md [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+## Build via GoLang directly
+
+go build -o ./build/bin/quai-cpu-miner main.go
```
Comment (gameofpointers) [inline README.md:17]: same as above

File: README.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+
+ZoneURLs: stores the URLs for the Zone chains. Should not be changed.
+
+Note that some of the values supplied in the config.yaml file can be overridden with the appropriate command and arguments.
+
+# Run the miner
+Setting the region and zone flags for mining location
+
+## Run via Makefile
+make region=0 zone=0 run-mine
```
Comment (gameofpointers) [inline README.md]: why reverse this order?

File: main.go [hunk 23]
```diff
@@ -34,28 +40,52 @@ type Miner struct {
...
+	latestId uint64
 }
 
-// Clients for RPC connection to the Prime, region, & zone node belonging to the
+// Clients for RPC connection to the Prime, region, & zone ports belonging to the
 // slice we are actively mining
 type SliceClients [common.HierarchyDepth]*ethclient.Client
 
-// getNodeClients takes in a config and retrieves the Prime, Region, and Zone client
+func connectToProxy(config util.Config) *util.MinerSession {
```
Comment (gameofpointers) [inline main.go:71]: maybe explain the function?

File: main.go [hunk 24]
```diff
@@ -34,28 +40,52 @@ type Miner struct {
...
+	// loc := config.Location
```
Comment (gameofpointers) [inline main.go]: comment

File: main.go [hunk 25]
```diff
@@ -116,37 +144,85 @@ func main() {
...
-	go m.resultLoop()
+	if config.Proxy {
+		m.proxyClient = connectToProxy(config)
+		go m.fetchPendingHeaderProxy()
+		go m.startProxyListener()
+		go m.subscribeProxy()
+	} else {
+		m.sliceClients = connectToSlice(config)
+		go m.fetchPendingHeaderNode()
+		// No separate call needed to start listeners.
```
Comment (gameofpointers) [inline main.go:165]: no need for this comment right

File: main.go [hunk 26]
```diff
@@ -246,13 +322,28 @@ func (m *Miner) hashratePrinter() {
...
 	for {
 		select {
 		case header := <-m.resultCh:
 			order, err := m.GetDifficultyOrder(header)
 			if err != nil {
+				// Ignore this block and go back to waiting for another one.
 				log.Println("Block mined has an invalid order")
+				continue
+			}
+			if !config.Proxy {
```
Comment (gameofpointers) [inline main.go]: can we make proxy as the flag in Miner struct? then no need to pass in the config to this function

File: util/miner_utils.go [hunk 27]
```diff
@@ -0,0 +1,107 @@
...
+	sync.Mutex
+	latestId uint64
+}
+
+const (
+	c_Max_Req_Size = 4096
+)
+
+func NewMinerConn(endpoint string) (*MinerSession, error) {
+	localaddr, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
```
Comment (gameofpointers) [inline util/miner_utils.go]: can we parametrize this IP here?

File: util/miner_utils.go [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+		log.Fatalf("Error: %v", err)
+		panic(err)
+	}
+
+	server, err := net.DialTCP("tcp", localaddr, remoteaddr)
+	if err != nil {
+		log.Fatalf("Error: %v", err)
+		panic(err)
+	}
+	server.SetDeadline(time.Time{})
```
Comment (gameofpointers) [inline util/miner_utils.go]: setting deadline with empty time input?

File: README.md (hunk 20 above)
Comment (Djadih) [reply README.md:11]: Done

File: README.md (hunk 21 above)
Comment (Djadih) [reply README.md:17]: Done

File: README.md (hunk 22 above)
Comment (Djadih) [reply README.md]: Reversed the reversal

File: main.go (hunk 23 above)
Comment (Djadih) [reply main.go:71]: Done

File: main.go (hunk 26 above)
Comment (Djadih) [reply main.go]: Good idea

File: main.go (hunk 25 above)
Comment (Djadih) [reply main.go:165]: Discussed irl, understood that this comment is there to provide symmetry with the proxy/non-proxy modes

File: util/miner_utils.go (hunk 27 above)
Comment (Djadih) [reply util/miner_utils.go]: Apparently you don't need to create a localaddr when you just want localhost and a random port. Made the change and deleted these lines.

File: util/miner_utils.go (hunk 28 above)
Comment (Djadih) [reply util/miner_utils.go]: This sets the connection to never timeout, but is default so no need to keep this line.
Removed.

**Review by wizeguyy — APPROVED** [review]

**Review by gameofpointers — APPROVED** [review]
